sympy/assumptions/ask.py

```diff
@@ -9,66 +9,1114 @@
 from sympy.assumptions.assume import (global_assumptions, Predicate,
         AppliedPredicate)
 from sympy.core.decorators import deprecated
-from sympy.utilities.decorator import classproperty, ClassPropertyDescriptor
-
-
-class Q(object):
-    """Supported ask keys."""
-    antihermitian = Predicate('antihermitian')
-    finite = Predicate('finite')
-    commutative = Predicate('commutative')
-    complex = Predicate('complex')
-    composite = Predicate('composite')
-    even = Predicate('even')
-    extended_real = Predicate('extended_real')
-    hermitian = Predicate('hermitian')
-    imaginary = Predicate('imaginary')
-    infinitesimal = Predicate('infinitesimal')
-    infinite = Predicate('infinite')
-    integer = Predicate('integer')
-    irrational = Predicate('irrational')
-    rational = Predicate('rational')
-    algebraic = Predicate('algebraic')
-    transcendental = Predicate('transcendental')
-    negative = Predicate('negative')
-    nonzero = Predicate('nonzero')
-    positive = Predicate('positive')
-    prime = Predicate('prime')
-    real = Predicate('real')
-    odd = Predicate('odd')
-    is_true = Predicate('is_true')
-    nonpositive = Predicate('nonpositive')
-    nonnegative = Predicate('nonnegative')
-    zero = Predicate('zero')
-
-    symmetric = Predicate('symmetric')
-    invertible = Predicate('invertible')
-    singular = Predicate('singular')
-    orthogonal = Predicate('orthogonal')
-    unitary = Predicate('unitary')
-    normal = Predicate('normal')
-    positive_definite = Predicate('positive_definite')
-    upper_triangular = Predicate('upper_triangular')
-    lower_triangular = Predicate('lower_triangular')
-    diagonal = Predicate('diagonal')
-    triangular = Predicate('triangular')
-    unit_triangular = Predicate('unit_triangular')
-    fullrank = Predicate('fullrank')
-    square = Predicate('square')
-    real_elements = Predicate('real_elements')
-    complex_elements = Predicate('complex_elements')
-    integer_elements = Predicate('integer_elements')
-
-    @classproperty
+from sympy.utilities.decorator import memoize_property
+
+
+# Deprecated predicates should be added to this list
+deprecated_predicates = [
+    'bounded',
+    'infinity'
+]
+
+# Memoization storage for predicates
+predicate_storage = {}
+
```

Comment:
I just mean something like

```
predicate_memo = memoize_property(predicate_storage)
```

here, and then use

```
@predicate_memo
```

everywhere below. It should speed up the import time, since the way it is now several decorators are created when the class is defined.